emcc.py

```diff
@@ -1761,9 +1755,28 @@ def include_and_export(name):
         '_asan_c_store_f', '_asan_c_store_d',
       ]
 
-      shared.Settings.GLOBAL_BASE = shared.Settings.ASAN_SHADOW_SIZE
-      shared.Settings.INITIAL_MEMORY += shared.Settings.ASAN_SHADOW_SIZE
-      assert shared.Settings.INITIAL_MEMORY < 2**32
+      if shared.Settings.ASAN_SHADOW_SIZE != -1:
+        diagnostics.warning('emcc', 'ASAN_SHADOW_SIZE is ignored and will be removed in a future release')
+
+      if shared.Settings.GLOBAL_BASE != -1:
+        exit_with_error("ASan does not support custom GLOBAL_BASE")
+
+      max_mem = shared.Settings.INITIAL_MEMORY
+      if shared.Settings.ALLOW_MEMORY_GROWTH:
+        max_mem = shared.Settings.MAXIMUM_MEMORY
+        if max_mem == -1:
+          exit_with_error('ASan requires a finite MAXIMUM_MEMORY')
+
+      shadow_size = max_mem // 8
+      shared.Settings.GLOBAL_BASE = shadow_size
+      shared.Settings.INITIAL_MEMORY += shadow_size
+      if shared.Settings.ALLOW_MEMORY_GROWTH:
+        shared.Settings.MAXIMUM_MEMORY += shadow_size
```

Comment:
Are these changes needed? It seems like normally the user-available memory is `[1024, max]` After you raise `GLOBAL_BASE`, it would be `[~max/8, max]`. The user has lost around `1/8` of available memory, but that seems ok? They can always raise `MAXIMUM_MEMORY`. Similar things can happen when switching between say dlmalloc and emmalloc (as dlmalloc has more efficient metadata usage). So it could be simpler to just bump `GLOBAL_BASE` and that's it unless I'm missing something.

(I guess some of the shadow memory would not be used - since available memory is slightly less than a multiple of 8 - but that seems fine too?)

Reply:
I wanted to spare the user from having to use different memory settings in their ASan builds and also avoid the situation where the user gets a hard error because the calculated shadow size is greater than their initial size. So I would say these changes aren't necessary, but they were deliberately made with usability in mind.

Reply:
The user can always increase memory size in both normal and ASan builds (so that it is enough for ASan, and slightly more than needed normally).

But how can that hard error happen? I'm missing something.

Reply:
That would happen if initial_memory < maximum_memory / 8. Shadow memory has to be statically allocated up front and can't grow.

Wouldn't it be better to keep the production build memory lower, though?

Reply:
Oh, I see. Yeah, that's a potential annoyance, I agree. But at least it would be a clear error.

Yes, production builds matter more. I guess it seems unlikely to me for someone to want to force the same amount in both. There is no precise guarantee of how much memory is available with malloc anyhow due to metadata, so it's all ballpark-esque.

Another reason I'd prefer to not change these values is that it can be surprising if the user asks for X MB and then runs the ASan build and sees it is using more. It seems simpler and less confusing to not change values silently.

But I don't feel very strongly, if I'm the only one that prefers to not change them then let's do the other way.

Reply:
(Anyhow in the long term multimemory is the real solution here.)

Reply:
Yeah I totally see the arguments for doing it a different way as well. @sbc100, do you have an opinion on which way is more user-friendly?